From 8aa572ba23d6ee2ef378fe1be190187c45e782ca Mon Sep 17 00:00:00 2001
From: Jomar Milan <jomarm@jomarm.com>
Date: Sun, 25 May 2025 19:27:48 -0700
Subject: [PATCH 1/2] maintainers: add MacaylaMarvelous81

---
 maintainers/maintainer-list.nix | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/maintainers/maintainer-list.nix b/maintainers/maintainer-list.nix
index 2a36e3741dd20..1be9ad9ed7622 100644
--- a/maintainers/maintainer-list.nix
+++ b/maintainers/maintainer-list.nix
@@ -15680,6 +15680,13 @@
     githubId = 401263;
     keys = [ { fingerprint = "1147 43F1 E707 6F3E 6F4B  2C96 B9A8 B592 F126 F8E8"; } ];
   };
+  MacaylaMarvelous81 = {
+    name = "Jomar Milan";
+    email = "jomarm@jomarm.com";
+    github = "MacaylaMarvelous81";
+    githubId = 29156241;
+    keys = [ { fingerprint = "F954 C5C9 5AE7 A312 183D  A76C 6AC4 6A6F 9A56 18D8"; } ];
+  };
   macronova = {
     name = "Sicheng Pan";
     email = "trivial@invariantspace.com";

From 80e24c8570c6d175d9e43f91ccd0a37018763b30 Mon Sep 17 00:00:00 2001
From: Jomar Milan <jomarm@jomarm.com>
Date: Mon, 26 May 2025 11:09:44 -0700
Subject: [PATCH 2/2] wasm4: init at 2.7.1

---
 .../0001-vendor-web-devtools-elsewhere.patch  |  11 ++
 ...2-fix-build-wasm3-by-disabling-debug.patch |  31 +++++
 pkgs/by-name/wa/wasm4/package.nix             | 126 ++++++++++++++++++
 3 files changed, 168 insertions(+)
 create mode 100644 pkgs/by-name/wa/wasm4/0001-vendor-web-devtools-elsewhere.patch
 create mode 100644 pkgs/by-name/wa/wasm4/0002-fix-build-wasm3-by-disabling-debug.patch
 create mode 100644 pkgs/by-name/wa/wasm4/package.nix

diff --git a/pkgs/by-name/wa/wasm4/0001-vendor-web-devtools-elsewhere.patch b/pkgs/by-name/wa/wasm4/0001-vendor-web-devtools-elsewhere.patch
new file mode 100644
index 0000000000000..debd8888e2736
--- /dev/null
+++ b/pkgs/by-name/wa/wasm4/0001-vendor-web-devtools-elsewhere.patch
@@ -0,0 +1,11 @@
+diff -ura web.a/package.json web.b/package.json
+--- web.a/package.json	2025-05-25 19:44:59.060428306 -0700
++++ web.b/package.json	2025-05-25 19:44:52.834495715 -0700
+@@ -16,7 +16,6 @@
+     "lint": "eslint src --fix"
+   },
+   "dependencies": {
+-    "@wasm4/web-devtools": "file:../../devtools/web",
+     "lit": "^3.1.2"
+   },
+   "devDependencies": {
diff --git a/pkgs/by-name/wa/wasm4/0002-fix-build-wasm3-by-disabling-debug.patch b/pkgs/by-name/wa/wasm4/0002-fix-build-wasm3-by-disabling-debug.patch
new file mode 100644
index 0000000000000..7b38126cee71d
--- /dev/null
+++ b/pkgs/by-name/wa/wasm4/0002-fix-build-wasm3-by-disabling-debug.patch
@@ -0,0 +1,31 @@
+diff -ura native.a/vendor/wasm3/CMakeLists.txt native.b/vendor/wasm3/CMakeLists.txt
+--- native.a/vendor/wasm3/CMakeLists.txt	2025-12-06 14:49:28.961968261 -0800
++++ native.b/vendor/wasm3/CMakeLists.txt	2025-12-06 15:00:16.322022055 -0800
+@@ -99,7 +99,7 @@
+ 
+ #-fno-optimize-sibling-calls
+ 
+-set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DDEBUG=1")
++#set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DDEBUG=1")
+ 
+ if(EMSCRIPTEN OR EMSCRIPTEN_LIB)
+ 
+diff -ura native.a/vendor/wasm3/source/m3_config_platforms.h native.b/vendor/wasm3/source/m3_config_platforms.h
+--- native.a/vendor/wasm3/source/m3_config_platforms.h	2025-12-06 14:49:04.189291630 -0800
++++ native.b/vendor/wasm3/source/m3_config_platforms.h	2025-12-06 15:09:41.626372007 -0800
+@@ -118,6 +118,7 @@
+  * Apply settings
+  */
+ 
++/*
+ # if defined (M3_COMPILER_MSVC)
+ #   define vectorcall   // For MSVC, better not to specify any call convention
+ # elif defined(__x86_64__)
+@@ -141,6 +142,7 @@
+ # elif defined (FOMU)
+ #   define vectorcall   __attribute__((section(".ramtext")))
+ # endif
++*/
+ 
+ #ifndef vectorcall
+ #define vectorcall
diff --git a/pkgs/by-name/wa/wasm4/package.nix b/pkgs/by-name/wa/wasm4/package.nix
new file mode 100644
index 0000000000000..ff7ea6e5b8007
--- /dev/null
+++ b/pkgs/by-name/wa/wasm4/package.nix
@@ -0,0 +1,126 @@
+{
+  lib,
+  fetchFromGitHub,
+  buildNpmPackage,
+  stdenv,
+  cmake,
+  xorg,
+  libxkbcommon,
+  alsa-lib,
+  libpulseaudio,
+  makeWrapper,
+  nix-update-script,
+}:
+let
+  version = "2.7.1";
+  src = fetchFromGitHub {
+    owner = "aduros";
+    repo = "wasm4";
+    tag = "v${version}";
+    hash = "sha256-QmSqQh+8w8pxoQcgHrsTfv/lIb5Xh7LkATnSMDUwjKo=";
+    # native runtime has submodule-vendored dependencies
+    fetchSubmodules = true;
+  };
+  webDevtools = buildNpmPackage (finalAttrs: {
+    pname = "web-devtools";
+    inherit version;
+
+    inherit src;
+    sourceRoot = "${src.name}/devtools/web";
+
+    npmDepsHash = "sha256-XPSnhQLdzIKY2qCiIfTIXQ8rOm7Tg5nPW7XEE1gmxRo=";
+    # dontNpmBuild = true;
+  });
+  webRuntime = buildNpmPackage (finalAttrs: {
+    pname = "wasm4-runtime";
+    inherit version;
+
+    inherit src;
+    sourceRoot = "${src.name}/runtimes/web";
+
+    patches = [ ./0001-vendor-web-devtools-elsewhere.patch ];
+    preBuildPhases = [ "provideWebDevtoolsPhase" ];
+    provideWebDevtoolsPhase = ''
+      mkdir -p node_modules/@wasm4
+      ln -s ${webDevtools}/lib/node_modules/@wasm4/web-devtools node_modules/@wasm4/web-devtools
+    '';
+    installPhase = ''
+      cp -r dist $out
+    '';
+
+    npmDepsHash = "sha256-LHyfNk0bCsWOB3uAIDqxyN7mpQJ6IsM5Oqx8V3/iNzc=";
+  });
+  nativeRuntime = stdenv.mkDerivation (finalAttrs: {
+    pname = "wasm4-runtime-native";
+    inherit version;
+
+    inherit src;
+    sourceRoot = "${finalAttrs.src.name}/runtimes/native";
+
+    # Without this patch, the vendored dependency wasm3 fails to build on newer clang versions.
+    patches = [ ./0002-fix-build-wasm3-by-disabling-debug.patch ];
+    cmakeFlags = [ "-DWASM_BACKEND=wasm3" ];
+    installPhase = ''
+      # this fails... but it does so on upstream ci too, and it just ignores it
+      # and uses the partial install.
+      cmake --install . --prefix $out --config Release --strip || true
+    '';
+    preFixupPhases = lib.optionals stdenv.hostPlatform.isLinux [ "wrapRuntimePhase" ];
+    wrapRuntimePhase = ''
+      wrapProgram $out/bin/wasm4 --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ libpulseaudio ]}
+    '';
+
+    nativeBuildInputs = [
+      cmake
+      makeWrapper
+    ];
+    buildInputs = lib.optionals stdenv.hostPlatform.isLinux [
+      xorg.libXrandr
+      xorg.libXinerama
+      xorg.libXcursor
+      xorg.libXi
+      xorg.libXext
+      libxkbcommon
+      alsa-lib
+      libpulseaudio
+    ];
+  });
+in
+buildNpmPackage (finalAttrs: {
+  pname = "wasm4";
+  inherit version;
+
+  inherit src;
+  sourceRoot = "${src.name}/cli";
+
+  preBuildPhases = [
+    "provideWebRuntimePhase"
+  ]
+  ++ lib.optionals stdenv.hostPlatform.isLinux [ "provideNativeLinuxRuntimePhase" ]
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [ "provideNativeDarwinRuntimePhase" ];
+  provideWebRuntimePhase = ''
+    rm assets/runtime
+    cp -r ${webRuntime} assets/runtime
+  '';
+  provideNativeLinuxRuntimePhase = ''
+    mkdir -p assets/natives
+    cp ${nativeRuntime}/bin/wasm4 assets/natives/wasm4-linux
+  '';
+  provideNativeDarwinRuntimePhase = ''
+    mkdir -p assets/natives
+    cp ${nativeRuntime}/bin/wasm4 assets/natives/wasm4-mac
+  '';
+
+  npmDepsHash = "sha256-p+sh1BHHmOY9mvUDOA/9KgoWTXwBfrlK2tqhXLQSeXk=";
+  dontNpmBuild = true;
+
+  meta = {
+    description = "Build retro games using WebAssembly for a fantasy console";
+    homepage = "https://wasm4.org";
+    downloadPage = "https://wasm4.org/docs/getting-started/setup";
+    license = lib.licenses.isc;
+    maintainers = with lib.maintainers; [ MacaylaMarvelous81 ];
+    mainProgram = "w4";
+    platforms = lib.platforms.all;
+  };
+})
